elki/src/main/java/de/lmu/ifi/dbs/elki/algorithm/associationrulemining/AssociationRuleGeneration.java

```diff
@@ -0,0 +1,226 @@
+   * @param interestMeasure Interestingness measure
+   * @param minmeasure Minimum threshold for interestingness measure
+   */
+  public AssociationRuleGeneration(AbstractFrequentItemsetAlgorithm frequentItemAlgo, AbstractInterestingnessMeasure interestMeasure, double minmeasure) {
+    super(frequentItemAlgo, interestMeasure, minmeasure);
+  }
+  
+  
+  
+  public AssociationRuleResult run(Database database, final Relation<BitVector> relation) {
+    
+    // Total number of transactions in transaction database;
+    final int totalTransactions = relation.size();
+    
+    // Run frequent itemset mining
+    FrequentItemsetsResult frequentResult = frequentItemAlgo.run(database);
+    
+    // Generate HashMap of resulting ArrayList with items in reverse order
+    LinkedHashMap<Itemset, Itemset> frequentItems = frequentResult.getRevHashMapRep();
+    
+    // Meta data for item tags
+    VectorFieldTypeInformation<BitVector> meta = frequentResult.getMeta();
+    
+    // Output List
+    List<AssociationRule> rules = new ArrayList<AssociationRule>();
+    
+    // Iterate through all frequent itemsets
+    for(Itemset itemset : frequentItems.keySet()) {
+      int[] itemsetSparse = itemset.toSparseRep();
+      
+      // stops if OneItemsets are reached
+      if (itemsetSparse.length > 1) {
+        List<Itemset> subsets = new ArrayList<Itemset>();
+        boolean[] pruning = new boolean[((int) Math.pow(2, itemsetSparse.length)) - 2];
+        calcSubsets(itemsetSparse, subsets, pruning, frequentItems);
+      
+        // Iterate through subsets
+        for(int i = 0; i < pruning.length; i++) {
+          if (pruning[i] == true) {
+            Itemset antecedent = subsets.get(i);
+            Itemset consequent = subsets.get(subsets.size() - i - 1);
+            double measure = this.interestMeasure.measure(totalTransactions, antecedent.getSupport(), consequent.getSupport(), itemset.getSupport());
+            
+            if(measure >= minmeasure && measure <= this.maxmeasure) {
+              rules.add(new AssociationRule(itemset, consequent, antecedent.getSupport(), measure));
+            } else {
+              if (this.interestMeasure instanceof Confidence) {
+                prune(antecedent,pruning, subsets);
+              }
+            }
+          }
+        }
+      } else {
+        break;
+      }
+    }
+    return new AssociationRuleResult("AssociationRule-Generation", "arule-generation", rules, meta);
+  }
+  
+  /**
+   * Fills a list with all subsets of a given itemset.
+   * 
+   * @author Frederic Sautter
+   *
+   * @param itemset Sparse representation of an itemset.
+   * @param subsets List to fill with subsets
+   * @param pruning Array for pruning
+   * @param map HashMap where the itemsets are stored
+   */
+  private static void calcSubsets(int[] itemset, List<Itemset> subsets, boolean[] pruning, HashMap<Itemset, Itemset> map) {
+    long creator = (long) Math.pow(2, itemset.length) - 2;
```

Comment:
Use integer math (bit shifting = power of two) rather than floating point. Why `-2`?